File: java/SlidingWindow/lengthOfLongestSubstring/lengthOfLongestSubstring.java
import java.util.*;

class lengthOfLongestSubstring {

    public int lengthOfLongestSubstringFunc(String s) {
        if (s.length() == 0) {
            return 0;
        }
        if (s.length() == 1) {
            return 1;
        }

        int left = 0;
        int maxLen = 0;

        Map<Character, Integer> lastIndex = new HashMap<>();

        for (int right = 0; right < s.length(); right++) {
            Character c = s.charAt(right);

            if (lastIndex.containsKey(c) && lastIndex.get(c) >= left) {
                left = lastIndex.get(c) + 1;
            }

            lastIndex.put(c, right);

            maxLen = Math.max(maxLen, right - left + 1);
        }

        return maxLen;
    }

}

File: cpp/SlidingWindow/lengthOfLongestSubstring/lengthOfLongestSubstring.cpp
#include <string>
#include <unordered_map>
class soultion {
  public:
    int lengthOfLongestSubstring(std::string str) {

        if (str.length() == 0)
            return 0;
        if (str.length() == 1)
            return 1;

        int left = 0;
        int maxLen = 0;

        std::unordered_map<char, int> lastIndex;

        for (int right = 0; right < str.size(); right++) {
            char c = str[right];
            if (lastIndex.count(c) && lastIndex[c] >= left) {
                left = lastIndex[c] + 1;
            }

            lastIndex[c] = right;
            maxLen = std::max(maxLen, right - left + 1);
        }
        return maxLen;
    }
};
